providers/redis/provider_legacy.go

```diff
@@ -0,0 +1,236 @@
+}
+
+// NewFailover returns a new redis provider using sentinel to determine the redis server to connect to.
+func NewFailover(cfg FailoverConfig) (*Provider, error) {
+	if cfg.MasterName == "" {
+		return nil, errConfigMasterNameEmpty
+	}
+
+	if cfg.Logger != nil {
+		redis.SetLogger(cfg.Logger)
+	}
+
+	db := redis.NewFailoverClient(&redis.FailoverOptions{
+		MasterName:         cfg.MasterName,
+		SentinelAddrs:      cfg.SentinelAddrs,
+		SentinelUsername:   cfg.SentinelUsername,
+		SentinelPassword:   cfg.SentinelPassword,
+		SlaveOnly:          cfg.SlaveOnly,
+		Username:           cfg.Username,
+		Password:           cfg.Password,
+		DB:                 cfg.DB,
+		MaxRetries:         cfg.MaxRetries,
+		MinRetryBackoff:    cfg.MinRetryBackoff,
+		MaxRetryBackoff:    cfg.MaxRetryBackoff,
+		DialTimeout:        cfg.DialTimeout,
+		ReadTimeout:        cfg.ReadTimeout,
+		WriteTimeout:       cfg.WriteTimeout,
+		PoolSize:           cfg.PoolSize,
+		MinIdleConns:       cfg.MinIdleConns,
+		MaxConnAge:         cfg.MaxConnAge,
+		PoolTimeout:        cfg.PoolTimeout,
+		IdleTimeout:        cfg.IdleTimeout,
+		IdleCheckFrequency: cfg.IdleCheckFrequency,
+		TLSConfig:          cfg.TLSConfig,
+	})
+
+	if err := db.Ping(context.Background()).Err(); err != nil {
+		return nil, errRedisConnection(err)
+	}
+
+	p := &Provider{
+		keyPrefix: cfg.KeyPrefix,
+		db:        db,
+	}
+
+	return p, nil
+}
+
+// NewFailoverCluster returns a new redis provider using a group of sentinels to determine the redis server to connect to.
+func NewFailoverCluster(cfg FailoverConfig) (*Provider, error) {
+	if cfg.MasterName == "" {
+		return nil, errConfigMasterNameEmpty
+	}
+
+	if cfg.Logger != nil {
+		redis.SetLogger(cfg.Logger)
+	}
+
+	db := redis.NewFailoverClusterClient(&redis.FailoverOptions{
+		MasterName:         cfg.MasterName,
+		SentinelAddrs:      cfg.SentinelAddrs,
+		SentinelUsername:   cfg.SentinelUsername,
+		SentinelPassword:   cfg.SentinelPassword,
+		RouteByLatency:     cfg.RouteByLatency,
+		RouteRandomly:      cfg.RouteRandomly,
+		SlaveOnly:          cfg.SlaveOnly,
+		Username:           cfg.Username,
+		Password:           cfg.Password,
+		DB:                 cfg.DB,
+		MaxRetries:         cfg.MaxRetries,
+		MinRetryBackoff:    cfg.MinRetryBackoff,
+		MaxRetryBackoff:    cfg.MaxRetryBackoff,
+		DialTimeout:        cfg.DialTimeout,
+		ReadTimeout:        cfg.ReadTimeout,
+		WriteTimeout:       cfg.WriteTimeout,
+		PoolSize:           cfg.PoolSize,
+		MinIdleConns:       cfg.MinIdleConns,
+		MaxConnAge:         cfg.MaxConnAge,
+		PoolTimeout:        cfg.PoolTimeout,
+		IdleTimeout:        cfg.IdleTimeout,
+		IdleCheckFrequency: cfg.IdleCheckFrequency,
+		TLSConfig:          cfg.TLSConfig,
+	})
+
+	if err := db.Ping(context.Background()).Err(); err != nil {
+		return nil, errRedisConnection(err)
+	}
+
+	p := &Provider{
+		keyPrefix: cfg.KeyPrefix,
+		db:        db,
+	}
+
+	return p, nil
+}
+
+func (p *Provider) getRedisSessionKey(sessionID []byte) string {
```

Comment:
Can you move the all the following functions to `provider_base.go` file without build tags to reduce the duplicated code, please? Because all of this functions are not involved by the redis lib version, just the constructors.

Reply:
This has been done for all appropriate functions.